File: _extensions/interlinks/interlinks.py
import itertools
import json
import panflute as pf

from quartodoc.interlinks import Ref, RefSyntaxError
from pathlib import Path
from plum import dispatch


# Hold all inventory items in a singleton -------------------------------------

# TODO: make entries into dataclass
# has fields: name, domain, role, priority, invname, full_uri


class InvLookupError(Exception):
    pass


class Inventories:
    def __init__(self):
        self.registry = {}

    def items(self):
        return itertools.chain(*self.registry.values())

    def load_inventory(self, inventory, url, invname):
        all_items = []
        for item in inventory["items"]:
            # TODO: what are the rules for inventories with overlapping names?
            #       it seems like this is where priority and using source name as an
            #       optional prefix in references is useful (e.g. siuba:a.b.c).
            full_uri = url + item["uri"].replace("$", item["name"])
            enh_item = {**item, "invname": invname, "full_uri": full_uri}
            all_items.append(enh_item)

        self.registry[invname] = all_items

    def lookup_reference(self, ref: Ref):
        # return global_inventory[ref]

        crnt_items = self.items()
        for field in ["name", "role", "domain", "invname"]:
            if field == "name":
                # target may have ~ option in front, so we strip it off
                field_value = ref.target.lstrip("~")
            else:
                field_value = getattr(ref, field)

            if field == "role":
                # for some reason, things like :func: are short for :function:.
                field_value = self.normalize_role(field_value)

            crnt_items = _filter_by_field(crnt_items, field, field_value)

        results = list(crnt_items)
        if not results:
            raise InvLookupError(
                f"Cross reference not found in an inventory file: `{ref}`"
            )

        if len(results) > 1:
            raise InvLookupError(
                f"Cross reference matches multiple entries.\n"
                f"Matching entries: {len(results)}\n"
                f"Reference: {ref}\n"
                f"Top 2 matches: \n  * {results[0]}\n  * {results[1]}"
            )

        return results[0]

    def normalize_role(self, role_name):
        if role_name == "func":
            return "function"

        return role_name


global_inventory = Inventories()


# Utility functions -----------------------------------------------------------


class ConfigError(Exception):
    pass


def get_path_to_root():
    # I have no idea how to get the documentation root,
    # except to get the path the extension script, which
    # lives in <root>/_extensions/interlinks, and work back
    p_root = Path(__file__).parent.parent.parent

    if p_root.name == "_extensions":
        return p_root.parent

    return p_root


def load_inventories(interlinks: dict):
    p_root = get_path_to_root()

    sources = interlinks["sources"]
    cache = interlinks.get("cache", "_inv")

    # load this sites inventory ----
    site_inv = interlinks.get("site_inv", "objects.json")

    json_data = json.load(open(p_root / site_inv))
    global_inventory.load_inventory(json_data, url="/", invname="")

    # load other inventories ----
    for doc_name, cfg in sources.items():

        fname = doc_name + "_objects.json"
        inv_path = p_root / Path(cache) / fname

        json_data = json.load(open(inv_path))

        global_inventory.load_inventory(json_data, url=cfg["url"], invname=doc_name)


def _filter_by_field(items, field_name: str, value: "str | None" = None):
    if value is None:
        return items

    return (item for item in items if item[field_name] == value)


def ref_to_anchor(raw: str, text: "str | pf.ListContainer | None"):
    """Return a Link element based on ref in interlink format

    Parameters
    ----------
    ref:
        The interlink reference (e.g. "my_module.my_function").
    text:
        The text to be displayed for the link.

    Examples
    --------

    >>> url = "https://example.org/functools.partial.html"
    >>> load_mock_inventory({"functools.partial": {"full_uri": url, "name": "functools.partial"}})
    >>> ref_to_anchor("functools.partial")
    Link(Str(functools.partial); url='https://example.org/functools.partial.html')

    >>> ref_to_anchor("~functools.partial")
    Link(Str(partial); url='https://example.org/functools.partial.html')
    """
    # TODO: for now we just mutate el

    try:
        ref = Ref.from_string(raw)
    except RefSyntaxError as e:
        pf.debug("WARNING: ", str(e))

    is_shortened = ref.target.startswith("~")

    entry = global_inventory.lookup_reference(ref)
    dst_url = entry["full_uri"]

    if not text:
        name = entry["name"] if entry["dispname"] == "-" else entry["dispname"]
        if is_shortened:
            # shorten names from module.sub_module.func_name -> func_name
            name = pf.Str(name.split(".")[-1])
        else:
            name = pf.Str(name)
    else:
        # when the element is an Link, content is a ListContainer, but it has to be
        # *splatted back into Link?
        if isinstance(text, pf.ListContainer):
            return pf.Link(*text, url=dst_url)
        elif isinstance(text, str):
            return pf.Link(pf.Str(text), url=dst_url)
        else:
            raise TypeError(f"Unsupported type: {type(text)}")

    return pf.Link(name, url=dst_url)


def parse_rst_style_ref(full_text):
    """
    Returns
    -------
    tuple
        The parsed title (None if no title specified), and corresponding reference.
    """

    import re

    m = re.match(r"(?P<text>.+?)\<(?P<ref>[a-zA-Z\.\-: _]+)\>", full_text)
    if m is None:
        # TODO: print a warning or something
        return full_text, None

    text, ref = m.groups()

    return ref, text


# Visitor ================================================================================


def prepare(doc: pf.Doc):

    meta = doc.get_metadata()

    try:
        interlinks = meta["interlinks"]
    except KeyError:
        raise ConfigError(
            "No interlinks.sources field detected in your metadata."
            "Please add this to your header:\n\n"
            "interlinks:"
            "\n sources:"
            "\n    - <source_name>: {url: ..., inv: ..., fallback: ... }"
        )

    load_inventories(interlinks)

    return doc


@dispatch
def visit(el, doc):
    return el
    # raise TypeError(f"Unsupported type: {type(el)}")


# TODO: the syntax :ref:`target` is not trivial to implement. The pandoc AST
# often embeds it in a list of Plain with other elements. Currently, we only
# support the syntax inside of links.
#
# @dispatch
# def visit(el: pf.Plain, doc):
#     cont = el.content
#     if len(cont) == 2 and cont[0] == pf.Str(":ref:") and isinstance(cont[1], pf.Code):
#         _, code = el.content
#
#         ref, title = parse_rst_style_ref(code.text)
#
#         return pf.Plain(ref_to_anchor(ref, title))
#
#     return el


@dispatch
def visit(el: pf.Link, doc):
    url = el.url
    if (url.startswith("%60") or url.startswith(":")) and url.endswith("%60"):
        # Get URL ----
        try:
            return ref_to_anchor(url.replace("%60", "`"), el.content)
        except InvLookupError as e:
            pf.debug("WARNING: " + str(e))
            if el.content:
                # Assuming content is a ListContainer(Str(...))
                body = el.content[0].text
            else:
                body = url.replace("%60", "")
            return pf.Code(body)

    return el


def main(doc=None):
    return pf.run_filter(visit, prepare=prepare, doc=None)


if __name__ == "__main__":
    main()

File: _extensions/interlinks/interlinks.lua
local function read_inv_text(filename)
    -- read file
    local file = io.open(filename, "r")
    if file == nil then
        return nil
    end
    local str = file:read("a")
    file:close()


    local project = str:match("# Project: (%S+)")
    local version = str:match("# Version: (%S+)")

    local data = { project = project, version = version, items = {} }

    local ptn_data =
        "^" ..
        "(.-)%s+" ..     -- name
        "([%S:]-):" ..   -- domain
        "([%S]+)%s+" ..  -- role
        "(%-?%d+)%s+" .. -- priority
        "(%S*)%s+" ..    -- uri
        "(.-)\r?$"       -- dispname


    -- Iterate through each line in the file content
    for line in str:gmatch("[^\r\n]+") do
        if not line:match("^#") then
            -- Match each line against the pattern
            local name, domain, role, priority, uri, dispName = line:match(ptn_data)

            -- if name is nil, raise an error
            if name == nil then
                error("Error parsing line: " .. line)
            end

            data.items[#data.items + 1] = {
                name = name,
                domain = domain,
                role = role,
                priority = priority,
                uri = uri,
                dispName = dispName
            }
        end
    end
    return data
end

local function read_json(filename)
    local file = io.open(filename, "r")
    if file == nil then
        return nil
    end
    local str = file:read("a")
    file:close()

    local decoded = quarto.json.decode(str)
    return decoded
end

local function read_inv_text_or_json(base_name)
    local file = io.open(base_name .. ".txt", "r")
    if file then
        -- TODO: refactors so we don't just close the file immediately
        io.close(file)
        json = read_inv_text(base_name .. ".txt")
    else
        json = read_json(base_name .. ".json")
    end

    return json
end

-- each inventory has entries: project, version, items
local inventory = {}

local function lookup(search_object)
    local results = {}
    for _, inv in ipairs(inventory) do
        for _, item in ipairs(inv.items) do
            -- e.g. :external+<inv_name>:<domain>:<role>:`<name>`
            if item.inv_name and item.inv_name ~= search_object.inv_name then
                goto continue
            end

            if item.name ~= search_object.name then
                goto continue
            end

            if search_object.role and item.role ~= search_object.role then
                goto continue
            end

            if search_object.domain and item.domain ~= search_object.domain then
                goto continue
            else
                if search_object.domain or item.domain == "py" then
                    table.insert(results, item)
                end

                goto continue
            end

            ::continue::
        end
    end

    if #results == 1 then
        return results[1]
    end
    if #results > 1 then
        quarto.log.warning("Found multiple matches for " .. search_object.name .. ", using the first match.")
        return results[1]
    end
    if #results == 0 then
        quarto.log.warning("Found no matches for object:\n", search_object)
    end

    return nil
end

local function mysplit(inputstr, sep)
    if sep == nil then
        sep = "%s"
    end
    local t = {}
    for str in string.gmatch(inputstr, "([^" .. sep .. "]+)") do
        table.insert(t, str)
    end
    return t
end

local function normalize_role(role)
    if role == "func" then
        return "function"
    end
    return role
end

local function copy_replace(original, key, new_value)
    -- First create a copy of the table
    local copy = {}
    for k, v in pairs(original) do
        copy[k] = v
    end

    -- Then replace the specific value
    copy[key] = new_value

    return copy
end

local function prepend_aliases(aliases)
    -- if str up to first period starts with an alias, then
    -- replace it with the full name.
    -- For example, suppose we have the alias quartodoc -> qd
    -- e.g. qd.Auto -> quartodoc.Auto
    -- e.g. qda.Auto -> qda.Auto

    local new_inv = {}
    new_inv["project"] = "aliases"
    new_inv["version"] = "0.0.9999" -- I have not begun to think about version...
    new_inv["items"] = {}

    for full, alias in pairs(aliases) do
        for _, inv in ipairs(inventory) do
            for _, item in ipairs(inv.items) do
                if string.sub(item.name, 1, string.len(full) + 1) == (full .. ".") then
                    -- replace full .. "." with alias .. "."
                    local prefix
                    if not alias or alias == "" then
                        prefix = ""
                    else
                        -- TODO: ensure alias doesn't end with period
                        prefix = alias .. "."
                    end
                    local new_name = prefix .. string.sub(item.name, string.len(full) + 2)
                    table.insert(new_inv.items, copy_replace(item, "name", new_name))
                end
            end
        end
    end
    table.insert(inventory, new_inv)
end

local function build_search_object(str)
    local starts_with_colon = str:sub(1, 1) == ":"
    local search = {}
    if starts_with_colon then
        local t = mysplit(str, ":")
        if #t == 2 then
            -- e.g. :py:func:`my_func`
            search.role = normalize_role(t[1])
            search.name = t[2]:match("%%60(.*)%%60")
        elseif #t == 3 then
            -- e.g. :py:func:`my_func`
            search.domain = t[1]
            search.role = normalize_role(t[2])
            search.name = t[3]:match("%%60(.*)%%60")
        elseif #t == 4 then
            -- e.g. :ext+inv:py:func:`my_func`
            search.external = true

            search.inv_name = t[1]:match("external%+(.*)")
            search.domain = t[2]
            search.role = normalize_role(t[3])
            search.name = t[4]:match("%%60(.*)%%60")
        else
            quarto.log.warning("couldn't parse this link: " .. str)
            return {}
        end
    else
        search.name = str:match("%%60(.*)%%60")
    end

    if search.name == nil then
        quarto.log.warning("couldn't parse this link: " .. str)
        return {}
    end

    if search.name:sub(1, 1) == "~" then
        search.shortened = true
        search.name = search.name:sub(2, -1)
    end
    return search
end

local function report_broken_link(link, search_object, replacement)
    -- TODO: how to unescape html elements like [?
    return pandoc.Code(pandoc.utils.stringify(link.content))
end

function Link(link)
    -- do not process regular links ----
    if not link.target:match("%%60") then
        return link
    end

    -- lookup item ----
    local search = build_search_object(link.target)
    local item = lookup(search)

    -- determine replacement, used if no link text specified ----
    local original_text = pandoc.utils.stringify(link.content)
    local replacement = search.name
    if search.shortened then
        local t = mysplit(search.name, ".")
        replacement = t[#t]
    end

    -- set link text ----
    if original_text == "" and replacement ~= nil then
        link.content = pandoc.Code(replacement)
    end

    -- report broken links ----
    if item == nil then
        return report_broken_link(link, search)
    end
    link.target = item.uri:gsub("%$$", search.name)


    return link
end

function Code(code)
    if not autolink then
        return code
    end

    -- allow text to be simple function call
    -- e.g. my_func() -> my_func
    -- e.g. a.b.call() -> a.b.call
    local text
    if code.text:match("%(%s*%)") then
        text = code.text:gsub("%(%s*%)", "")
    else
        text = code.text
    end


    -- return code.attr
    local search = build_search_object("%60" .. text .. "%60")
    local item = lookup(search)

    -- determine replacement, used if no link text specified ----
    if item == nil then
        quarto.log.warning(code)
        return code
    end

    return pandoc.Link(code, item.uri:gsub("%$$", search.name))
end

local function fixup_json(json, prefix, attach)
    for _, item in ipairs(json.items) do
        item.uri = prefix .. item.uri
    end
    table.insert(inventory, json)
end

return {
    {
        Meta = function(meta)
            local json
            local prefix
            local aliases

            -- set globals from config
            if meta.interlinks and meta.interlinks.autolink then
                autolink = true
            else
                autolink = false
            end

            if meta.interlinks and meta.interlinks.aliases then
                aliases = meta.interlinks.aliases
            else
                aliases = {}
            end

            -- process sources
            if meta.interlinks and meta.interlinks.sources then
                for k, v in pairs(meta.interlinks.sources) do
                    local base_name = quarto.project.offset .. "/_inv/" .. k .. "_objects"
                    json = read_inv_text_or_json(base_name)
                    prefix = pandoc.utils.stringify(v.url)
                    if json ~= nil then
                        fixup_json(json, prefix)
                    end
                end
            end
            json = read_inv_text_or_json(quarto.project.offset .. "/objects")
            if json ~= nil then
                fixup_json(json, "/")
            end

            prepend_aliases(aliases)
        end
    },
    {
        Link = Link,
        Code = Code
    }
}
